File: parallel_visualizer/src/parallel/prefix_sum/parallel.py
import math
import time

def serial_prefix_sum(input_data):
    n = len(input_data)
    m = 2 ** math.ceil(math.log2(n))
    prefix_sum = [0] * m

    # Copy input to prefix_sum
    for i in range(n):
        prefix_sum[i] = input_data[i]

    # Up-sweep (reduction) phase
    d = 0
    while (1 << d) < m:
        stride = 1 << (d + 1)
        for i in range(stride - 1, m, stride):
            prefix_sum[i] += prefix_sum[i - (stride >> 1)]
        d += 1

    # Down-sweep phase
    prefix_sum[m - 1] = 0
    d -= 1
    while d >= 0:
        stride = 1 << (d + 1)
        for i in range(stride - 1, m, stride):
            t = prefix_sum[i - (stride >> 1)]
            prefix_sum[i - (stride >> 1)] = prefix_sum[i]
            prefix_sum[i] += t
        d -= 1

    return prefix_sum

def main():
    input_file = "small_input.txt"
    with open(input_file, "r") as f:
        n = int(f.readline().strip())
        data = list(map(int, f.readline().strip().split()))

    start_time = time.time()

    prefix_sum = serial_prefix_sum(data)

    end_time = time.time()

    duration = (end_time - start_time) * 1000
    print(f"{duration:.2f} ms")

    print(" ".join(map(str, prefix_sum[:n + 1])))

if __name__ == "__main__":
    main()

File: parallel_visualizer/src/parallel/prefix_sum/parallel.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <chrono>
#include <omp.h>
#include <iomanip>

using namespace std;
typedef long long ll;

// Blelloch scan algorithm
// don't forget to pad with zeros
vector<long long> parallel_prefix_sum(const vector<int> &input)
{
    int n = input.size();
    int m = pow(2, ceil(log2(n)));
    vector<long long> prefix_sum(m, 0);

#pragma omp parallel for
    for (int i = 0; i < n; ++i)
    {
        prefix_sum[i] = input[i];
    }

    // Up-sweep (reduction) phase
    for (int d = 0; (1 << d) < m; ++d)
    {
        int stride = 1 << (d + 1);
#pragma omp parallel for
        for (int i = stride - 1; i < m; i += stride)
        {
            if (i - (stride >> 1) >= 0)
            {
                prefix_sum[i] += prefix_sum[i - (stride >> 1)];
            }
        }
    }

    // Down-sweep phase
    prefix_sum[m - 1] = 0;
    for (int d = (int)log2(m) - 1; d >= 0; --d)
    {
        int stride = 1 << (d + 1);
#pragma omp parallel for
        for (int i = stride - 1; i < m; i += stride)
        {
            if (i - (stride >> 1) >= 0)
            {
                int t = prefix_sum[i - (stride >> 1)];
                prefix_sum[i - (stride >> 1)] = prefix_sum[i];
                prefix_sum[i] += t;
            }
        }
    }
    prefix_sum.resize(n + 1);

    return prefix_sum;
}

int main()
{
    freopen("small_input.txt", "r", stdin);
    int n;
    cin >> n;
    vector<int> v(n);
    for (int &x : v)
    {
        cin >> x;
    }

    auto start = chrono::high_resolution_clock::now();
    vector<long long> p = parallel_prefix_sum(v);
    auto end = chrono::high_resolution_clock::now();

    chrono::duration<double, milli> duration = end - start;
    cout << fixed << setprecision(2) << duration.count() << " ms\n";

    // This is the correct answer for small input
    // cout << "Expected output: 0 9 19 27 28 34 38 41 43 53 61\n";
    // cout << "Computed output: ";
    // for (int i = 0; i <= n; ++i)
    // {
    //     cout << p[i] << " ";
    // }
    // cout << "\n";

    return 0;
}
